composer/callbacks/early_stopper.py

```diff
@@ -0,0 +1,139 @@
+            least that amount. Defaults to 0.
+        patience (int | str | Time, optional): The interval of time the monitored metric can not improve without stopping
+            training. Defaults to 1 epoch.
+        
+    """
+
+    def __init__(
+        self,
+        monitor: str,
+        dataloader_label: str,
+        comp: Optional[Callable[[
+            Any,
+            Any,
+        ], bool]] = None,
+        min_delta: float = 0.0,
+        patience: Union[int, str, Time] = 1,
+    ):
+        self.monitor = monitor
+        self.dataloader_label = dataloader_label
+        self.comp = comp
+        self.min_delta = abs(min_delta)
+        if self.comp is None:
+            if any(substr in monitor.lower() for substr in ["loss", "error", "perplexity"]):
+                self.comp = np.less
+            else:
+                self.comp = np.greater
+
+        self.best = None
+        self.best_occurred = None
+
+        if isinstance(patience, str):
+            self.patience = Time.from_timestring(patience)
+        elif isinstance(patience, int):
+            self.patience = Time(patience, TimeUnit.EPOCH)
+        else:
+            self.patience = patience
+            if self.patience.unit not in (TimeUnit.EPOCH, TimeUnit.BATCH):
+                raise ValueError("If `patience` is an instance of Time, it must have units of EPOCH or BATCH.")
+
+    def _get_monitored_metric(self, state: State):
+        if self.dataloader_label in state.current_metrics:
+            if self.monitor in state.current_metrics[self.dataloader_label]:
+                return state.current_metrics[self.dataloader_label][self.monitor]
+        raise ValueError(f"Couldn't find the metric {self.monitor} with the dataloader label {self.dataloader_label}."
+                         "Check that the dataloader_label is set to 'eval', 'train' or the evaluator name.")
+
+    def _update_stopper_state(self, state: State):
+        metric_val = self._get_monitored_metric(state)
+
+        assert self.comp is not None
+        if self.best is None:
+            self.best = metric_val
+            self.best_occurred = state.timer.get_timestamp()
+        elif self.comp(metric_val, self.best) and abs(metric_val - self.best) > self.min_delta:
+            self.best = metric_val
+            self.best_occurred = state.timer.get_timestamp()
+
+        assert self.best_occurred is not None
+        if self.patience.unit == TimeUnit.EPOCH:
+            if state.timer.epoch - self.best_occurred.epoch > self.patience:
+                state.max_duration = state.timer.batch
+        elif self.patience.unit == TimeUnit.BATCH:
+            if state.timer.batch - self.best_occurred.batch > self.patience:
+                state.max_duration = state.timer.batch
+        else:
+            raise ValueError(f"The units of `patience` should be EPOCH or BATCH.")
+
+    def eval_end(self, state: State, logger: Logger) -> None:
+        if self.dataloader_label != "train":
```

Comment:
For these checks (and the ones below), is it possible to compare `self.dataloader_label` against `state.dataloader_label`? The dataloader label on the state will be set to the currently active dataloader.

(I don't think this check would work if using multiple evaluators, and the train dataloader is not guaranteed to be called 'train')

Reply:
I think the checks will work with multiple evaluators, unless they label one of their evaluators as "train". That test doesn't select for which dataloader metrics to use, it just tests whether the user is trying to get a training metric or an eval metric to decide which event to run on. Could you correct if I'm wrong?

In multiple places, docs/code say that the training dataloader will be labelled 'train', could you elaborate?